File: sdk/packages/node/iii/src/iii-types.ts
export enum MessageType {
  RegisterFunction = 'registerfunction',
  UnregisterFunction = 'unregisterfunction',
  RegisterService = 'registerservice',
  InvokeFunction = 'invokefunction',
  InvocationResult = 'invocationresult',
  RegisterTriggerType = 'registertriggertype',
  RegisterTrigger = 'registertrigger',
  UnregisterTrigger = 'unregistertrigger',
  UnregisterTriggerType = 'unregistertriggertype',
  TriggerRegistrationResult = 'triggerregistrationresult',
  WorkerRegistered = 'workerregistered',
}

export type RegisterTriggerTypeMessage = {
  message_type: MessageType.RegisterTriggerType
  id: string
  description: string
}

export type UnregisterTriggerTypeMessage = {
  message_type: MessageType.UnregisterTriggerType
  id: string
}

export type UnregisterTriggerMessage = {
  message_type: MessageType.UnregisterTrigger
  id: string
  type?: string
}

export type TriggerRegistrationResultMessage = {
  message_type: MessageType.TriggerRegistrationResult
  id: string
  type: string
  function_id: string
  result?: unknown
  error?: unknown
}

export type RegisterTriggerMessage = {
  message_type: MessageType.RegisterTrigger
  id: string
  type: string
  function_id: string
  config: unknown
}

export type RegisterServiceMessage = {
  message_type: MessageType.RegisterService
  id: string
  name?: string
  description?: string
  parent_service_id?: string
}

/**
 * Authentication configuration for HTTP-invoked functions.
 *
 * - `hmac` -- HMAC signature verification using a shared secret.
 * - `bearer` -- Bearer token authentication.
 * - `api_key` -- API key sent via a custom header.
 */
export type HttpAuthConfig =
  | { type: 'hmac'; secret_key: string }
  | { type: 'bearer'; token_key: string }
  | { type: 'api_key'; header: string; value_key: string }

/**
 * Configuration for registering an HTTP-invoked function (Lambda, Cloudflare
 * Workers, etc.) instead of a local handler.
 */
export type HttpInvocationConfig = {
  /** URL to invoke. */
  url: string
  /** HTTP method. Defaults to `POST`. */
  method?: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE'
  /** Timeout in milliseconds. */
  timeout_ms?: number
  /** Custom headers to send with the request. */
  headers?: Record<string, string>
  /** Authentication configuration. */
  auth?: HttpAuthConfig
}

export type RegisterFunctionFormat = {
  /**
   * The name of the parameter
   */
  name?: string
  /**
   * The description of the parameter
   */
  description?: string
  /**
   * The type of the parameter
   */
  type: 'string' | 'number' | 'boolean' | 'object' | 'array' | 'null' | 'map'
  /**
   * The body of the parameter
   */
  properties?: Record<string, RegisterFunctionFormat>
  /**
   * The items of the parameter
   */
  items?: RegisterFunctionFormat
  /**
   * Whether the parameter is required
   */
  required?: string[]
}

export type RegisterFunctionMessage = {
  message_type: MessageType.RegisterFunction
  /**
   * The path of the function (use :: for namespacing, e.g. external::my_lambda)
   */
  id: string
  /**
   * The description of the function
   */
  description?: string
  /**
   * The request format of the function
   */
  request_format?: RegisterFunctionFormat
  /**
   * The response format of the function
   */
  response_format?: RegisterFunctionFormat
  metadata?: Record<string, unknown>
  /**
   * HTTP invocation config for external HTTP functions (Lambda, Cloudflare Workers, etc.)
   */
  invocation?: HttpInvocationConfig
}

/**
 * Routing action for {@link TriggerRequest}. Determines how the engine
 * handles the invocation.
 *
 * - `enqueue` -- Routes through a named queue for async processing.
 * - `void` -- Fire-and-forget, no response.
 */
export type TriggerAction = { type: 'enqueue'; queue: string } | { type: 'void' }

/**
 * Result returned when a function is invoked with `TriggerAction.Enqueue`.
 */
export type EnqueueResult = {
  /** Unique receipt ID for the enqueued message. */
  messageReceiptId: string
}

/**
 * Request object passed to {@link ISdk.trigger}.
 *
 * @typeParam TInput - Type of the payload.
 */
export type TriggerRequest<TInput = unknown> = {
  /** ID of the function to invoke. */
  function_id: string
  /** Payload to pass to the function. */
  payload: TInput
  /** Routing action. Omit for synchronous request/response. */
  action?: TriggerAction
  /** Override the default invocation timeout in milliseconds. */
  timeoutMs?: number
}

export type InvokeFunctionMessage = {
  message_type: MessageType.InvokeFunction
  /**
   * This is optional for async invocations
   */
  invocation_id?: string
  /**
   * The path of the function
   */
  function_id: string
  /**
   * The data to pass to the function
   */
  data: unknown
  /**
   * W3C trace-context traceparent header for distributed tracing
   */
  traceparent?: string
  /**
   * W3C baggage header for cross-cutting context propagation
   */
  baggage?: string
  /**
   * Trigger action for queue routing or fire-and-forget
   */
  action?: TriggerAction
}

export type InvocationResultMessage = {
  message_type: MessageType.InvocationResult
  /**
   * The id of the invocation
   */
  invocation_id: string
  /**
   * The path of the function
   */
  function_id: string
  result?: unknown
  error?: unknown
  /**
   * W3C trace-context traceparent header for distributed tracing
   */
  traceparent?: string
  /**
   * W3C baggage header for cross-cutting context propagation
   */
  baggage?: string
}

/**
 * Metadata about a registered function, returned by `ISdk.listFunctions`.
 */
export type FunctionInfo = {
  /** Unique function identifier. */
  function_id: string
  /** Human-readable description. */
  description?: string
  /** Schema describing expected request format. */
  request_format?: RegisterFunctionFormat
  /** Schema describing expected response format. */
  response_format?: RegisterFunctionFormat
  /** Arbitrary metadata attached to the function. */
  metadata?: Record<string, unknown>
}

/**
 * Information about a registered trigger.
 */
export type TriggerInfo = {
  /** Unique trigger identifier. */
  id: string
  /** Type of the trigger (e.g. `http`, `cron`, `queue`). */
  trigger_type: string
  /** ID of the function this trigger is bound to. */
  function_id: string
  /** Trigger-specific configuration. */
  config?: unknown
}

/** Worker connection status. */
export type WorkerStatus = 'connected' | 'available' | 'busy' | 'disconnected'

/**
 * Metadata about a connected worker, returned by `ISdk.listWorkers`.
 */
export type WorkerInfo = {
  /** Unique worker identifier assigned by the engine. */
  id: string
  /** Display name of the worker. */
  name?: string
  /** Runtime environment (e.g. `node`, `python`, `rust`). */
  runtime?: string
  /** SDK version. */
  version?: string
  /** Operating system info. */
  os?: string
  /** IP address of the worker. */
  ip_address?: string
  /** Current connection status. */
  status: WorkerStatus
  /** Timestamp (ms since epoch) when the worker connected. */
  connected_at_ms: number
  /** Number of functions registered by this worker. */
  function_count: number
  /** List of function IDs registered by this worker. */
  functions: string[]
  /** Number of currently active invocations. */
  active_invocations: number
}

export type WorkerRegisteredMessage = {
  message_type: MessageType.WorkerRegistered
  worker_id: string
}

export type UnregisterFunctionMessage = {
  message_type: MessageType.UnregisterFunction
  id: string
}

/**
 * Serializable reference to one end of a streaming channel. Can be included
 * in invocation payloads to pass channel endpoints between workers.
 */
export type StreamChannelRef = {
  /** Unique channel identifier. */
  channel_id: string
  /** Access key for authentication. */
  access_key: string
  /** Whether this ref is for reading or writing. */
  direction: 'read' | 'write'
}

export type IIIMessage =
  | RegisterFunctionMessage
  | UnregisterFunctionMessage
  | InvokeFunctionMessage
  | InvocationResultMessage
  | RegisterServiceMessage
  | RegisterTriggerMessage
  | RegisterTriggerTypeMessage
  | UnregisterTriggerMessage
  | UnregisterTriggerTypeMessage
  | TriggerRegistrationResultMessage
  | WorkerRegisteredMessage

File: sdk/packages/python/iii/src/iii/iii_types.py
"""III message types."""

from enum import Enum
from typing import Any, Literal

from iii_helpers.http import HttpInvocationConfig
from pydantic import BaseModel, ConfigDict, Field


class MessageType(str, Enum):
    """Message types for iii communication."""

    REGISTER_FUNCTION = "registerfunction"
    UNREGISTER_FUNCTION = "unregisterfunction"
    REGISTER_SERVICE = "registerservice"
    INVOKE_FUNCTION = "invokefunction"
    INVOCATION_RESULT = "invocationresult"
    REGISTER_TRIGGER_TYPE = "registertriggertype"
    REGISTER_TRIGGER = "registertrigger"
    UNREGISTER_TRIGGER = "unregistertrigger"
    UNREGISTER_TRIGGER_TYPE = "unregistertriggertype"
    TRIGGER_REGISTRATION_RESULT = "triggerregistrationresult"
    WORKER_REGISTERED = "workerregistered"


class RegisterTriggerTypeMessage(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    description: str
    trigger_request_format: Any | None = Field(default=None)
    call_request_format: Any | None = Field(default=None)
    message_type: MessageType = Field(default=MessageType.REGISTER_TRIGGER_TYPE, alias="type")


class UnregisterTriggerTypeMessage(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    message_type: MessageType = Field(default=MessageType.UNREGISTER_TRIGGER_TYPE, alias="type")


class UnregisterTriggerMessage(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    message_type: MessageType = Field(default=MessageType.UNREGISTER_TRIGGER, alias="type")
    trigger_type: str | None = Field(default=None, alias="trigger_type")


class TriggerRegistrationResultMessage(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    trigger_type: str = Field(alias="trigger_type")
    function_id: str = Field()
    result: Any = None
    error: Any = None
    message_type: MessageType = Field(default=MessageType.TRIGGER_REGISTRATION_RESULT, alias="type")


class RegisterTriggerTypeInput(BaseModel):
    """Input for registering a trigger type.

    Attributes:
        id: Unique identifier for the trigger type.
        description: Human-readable description of the trigger type.
        trigger_request_format: JSON Schema describing the expected trigger config.
        call_request_format: JSON Schema describing the payload sent to functions.
    """

    id: str = Field(description="Unique identifier for the trigger type.")
    description: str = Field(description="Human-readable description of the trigger type.")
    trigger_request_format: Any | None = Field(
        default=None, description="JSON Schema for trigger configuration."
    )
    call_request_format: Any | None = Field(
        default=None, description="JSON Schema for the call request payload."
    )


class RegisterTriggerInput(BaseModel):
    """Input for registering a trigger (matches Node SDK's RegisterTriggerInput).

    Attributes:
        type: Trigger type identifier (e.g. ``http``, ``queue``, ``cron``).
        function_id: ID of the function this trigger invokes.
        config: Trigger-type-specific configuration.
        metadata: Arbitrary metadata attached to the trigger.
    """

    type: str = Field(description="Trigger type identifier.")
    function_id: str = Field(description="ID of the function this trigger invokes.")
    config: Any = Field(default=None, description="Trigger-type-specific configuration.")
    metadata: Any | None = Field(
        default=None, description="Arbitrary metadata attached to the trigger."
    )


class RegisterTriggerMessage(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    trigger_type: str = Field(alias="trigger_type")
    function_id: str = Field()
    config: Any
    metadata: Any | None = Field(default=None)
    message_type: MessageType = Field(default=MessageType.REGISTER_TRIGGER, alias="type")


class RegisterFunctionFormat(BaseModel):
    """Format definition for function parameters.

    Attributes:
        name: Parameter name.
        type: Type string (``string``, ``number``, ``boolean``, ``object``, ``array``, ``null``, ``map``).
        description: Human-readable description of the parameter.
        body: Nested fields for object types.
        items: Item schema for array types.
        required: Whether the parameter is required.
    """

    name: str
    type: str  # 'string' | 'number' | 'boolean' | 'object' | 'array' | 'null' | 'map'
    description: str | None = None
    body: list["RegisterFunctionFormat"] | None = None
    items: "RegisterFunctionFormat | None" = None
    required: bool = False


class RegisterFunctionInput(BaseModel):
    """Input for registering a function, matches Node.js RegisterFunctionInput.

    Attributes:
        id: Unique function identifier.
        description: Human-readable description.
        request_format: Schema describing expected input.
        response_format: Schema describing expected output.
        metadata: Arbitrary metadata attached to the function.
        invocation: HTTP invocation config for externally hosted functions.
    """

    id: str = Field(description="Unique function identifier.")
    description: str | None = Field(default=None, description="Human-readable description.")
    request_format: RegisterFunctionFormat | dict[str, Any] | None = Field(
        default=None, description="Schema describing expected input."
    )
    response_format: RegisterFunctionFormat | dict[str, Any] | None = Field(
        default=None, description="Schema describing expected output."
    )
    metadata: Any | None = Field(default=None, description="Arbitrary metadata attached to the function.")
    invocation: HttpInvocationConfig | None = Field(
        default=None,
        description="HTTP invocation config for externally hosted functions.",
    )


class RegisterFunctionMessage(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str = Field()
    description: str | None = None
    request_format: RegisterFunctionFormat | dict[str, Any] | None = Field(default=None)
    response_format: RegisterFunctionFormat | dict[str, Any] | None = Field(default=None)
    metadata: Any | None = None
    invocation: HttpInvocationConfig | None = None
    message_type: MessageType = Field(default=MessageType.REGISTER_FUNCTION, alias="type")


class TriggerActionEnqueue(BaseModel):
    """Routes the invocation through a named queue for async processing.

    Requires a queue worker in the project — run ``iii worker add queue``.
    Without it the trigger rejects with ``enqueue_error`` (no queue provider).

    Attributes:
        type: Always ``'enqueue'``.
        queue: Name of the target queue.
    """

    type: Literal["enqueue"] = "enqueue"
    queue: str


class TriggerActionVoid(BaseModel):
    """Fire-and-forget routing. No response is returned.

    Attributes:
        type: Always ``'void'``.
    """

    type: Literal["void"] = "void"


TriggerAction = TriggerActionEnqueue | TriggerActionVoid
"""Routing action for trigger requests."""


class MiddlewareFunctionInput(BaseModel):
    """Input passed to the RBAC middleware function on every function invocation
    through the RBAC port.

    Attributes:
        function_id: ID of the function being invoked.
        payload: Payload sent by the caller.
        action: Routing action, if any.
        context: Auth context returned by the auth function for this session.
    """

    function_id: str = Field(description="ID of the function being invoked.")
    payload: dict[str, Any] = Field(description="Payload sent by the caller.")
    action: TriggerActionEnqueue | TriggerActionVoid | None = Field(
        default=None, description="Routing action, if any.",
    )
    context: dict[str, Any] = Field(
        description="Auth context returned by the auth function for this session.",
    )


class TriggerRequest(BaseModel):
    """Request object for ``trigger()``.

    Attributes:
        function_id: ID of the function to invoke.
        payload: Data to pass to the function.
        action: Routing action, ``None`` for sync, ``TriggerAction.Enqueue(...)``
            for queue, ``TriggerAction.Void()`` for fire-and-forget.
        timeout_ms: Override the default invocation timeout.
        metadata: Arbitrary per-invocation metadata delivered to the handler
            as a separate channel (not folded into ``payload``).
    """

    function_id: str = Field(description="ID of the function to invoke.")
    payload: Any = Field(default=None, description="Data to pass to the function.")
    action: TriggerActionEnqueue | TriggerActionVoid | None = Field(
        default=None,
        description=(
            "Routing action: ``None`` for sync, "
            "``TriggerAction.Enqueue(...)`` for queue, "
            "``TriggerAction.Void()`` for fire-and-forget."
        ),
    )
    timeout_ms: int | None = Field(default=None, description="Override the default invocation timeout.")
    metadata: Any | None = Field(
        default=None,
        description="Arbitrary per-invocation metadata delivered to the handler.",
    )


class InvokeFunctionMessage(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    function_id: str = Field()
    data: Any
    metadata: Any | None = Field(default=None)
    invocation_id: str | None = Field(default=None)
    traceparent: str | None = Field(default=None)
    baggage: str | None = Field(default=None)
    action: TriggerActionEnqueue | TriggerActionVoid | None = Field(default=None)
    message_type: MessageType = Field(default=MessageType.INVOKE_FUNCTION, alias="type")


class InvocationResultMessage(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    invocation_id: str = Field()
    function_id: str = Field()
    result: Any = None
    error: Any = None
    traceparent: str | None = Field(default=None)
    baggage: str | None = Field(default=None)
    message_type: MessageType = Field(default=MessageType.INVOCATION_RESULT, alias="type")


class WorkerRegisteredMessage(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    worker_id: str = Field()
    message_type: MessageType = Field(default=MessageType.WORKER_REGISTERED, alias="type")


class UnregisterFunctionMessage(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    message_type: MessageType = Field(default=MessageType.UNREGISTER_FUNCTION, alias="type")


class StreamChannelRef(BaseModel):
    """Reference to a streaming channel for worker-to-worker data transfer.

    Attributes:
        channel_id: Unique channel identifier.
        access_key: Secret key for authenticating channel access.
        direction: Channel direction (``read`` or ``write``).
    """

    channel_id: str = Field(description="Unique channel identifier.")
    access_key: str = Field(description="Secret key for authenticating channel access.")
    direction: Literal["read", "write"] = Field(description="Channel direction (``reader`` or ``writer``).")


class OtelLogEvent(BaseModel):
    """OTEL log event received from the engine via ``on_log``."""

    timestamp_unix_nano: int
    observed_timestamp_unix_nano: int
    severity_number: int
    severity_text: str
    body: str
    attributes: dict[str, Any] = Field(default_factory=dict)
    trace_id: str | None = None
    span_id: str | None = None
    resource: dict[str, str] = Field(default_factory=dict)
    service_name: str = ""
    instrumentation_scope_name: str | None = None
    instrumentation_scope_version: str | None = None


IIIMessage = (
    RegisterFunctionMessage
    | UnregisterFunctionMessage
    | InvokeFunctionMessage
    | InvocationResultMessage
    | RegisterTriggerMessage
    | RegisterTriggerTypeMessage
    | UnregisterTriggerMessage
    | UnregisterTriggerTypeMessage
    | TriggerRegistrationResultMessage
    | WorkerRegisteredMessage
)
